### PR TITLE
feat!: add fail-guard to Goal

### PR DESCRIPTION
**Pull request by Laura-Berkers**
## Description

This PR adds a failure condition to the Goal class. The used can provide a fail-guard upon construction of the Goal. If no fail-guard is provided, the goal will never fail.

The fail-guard is checked in de UpdateStatus method of the goal, before the heuristic. Thus, failure takes precedence over success.

This PR also changed all usages of `System.Func<T, bool>` to `System.Predicate<T>`, which is more semantically accurate.

And this PR also changes the `GetStatus` method in `IGoal` to `UpdateStatus` to be in line with other classes/interfaces implementing the `ICompletable` interface.

Lastly, this PR removes all remaining imports of the plain `System` namespace from the project, to avoid collision between `System.Action` and `Aplib.Core.Intent.Actions.Action`.

## Breaking change

- Because a `System.Func<T, bool>` cannot be implicitly converted to a `System.Predicate<T>`.
- Because the `GetStatus` method is replaced by `UpdateStatus`.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch

### CODE REVIEW

File: Aplib.Core/Belief/Beliefs/Belief.cs [hunk 1]
```diff
@@ -1,6 +1,4 @@
-﻿using System;
```
Comment (JoenvandeVorle) [inline Aplib.Core/Belief/Beliefs/Belief.cs:1]: So why did you remove this everywhere? I'm not too much of a fan of it

**Comment by Laura-Berkers** [comment]
Updated PR description to mentions the removal of `System` imports.

File: Aplib.Core/Desire/Goals/Goal.cs [hunk 2]
```diff
@@ -51,70 +56,143 @@ public class Goal<TBeliefSet> : IGoal<TBeliefSet>, IDocumented
         /// </summary>
         public ITactic<TBeliefSet> Tactic { get; }
 
-        /// <inheritdoc />
+        /// <summary>
+        /// Gets the completion status of the goal.
+        /// This value may need to be updated first using the <see cref="UpdateStatus"/> method.
+        /// </summary>
+        /// <seealso cref="UpdateStatus"/>
         public CompletionStatus Status { get; protected set; }
```
Comment (JoenvandeVorle) [inline Aplib.Core/Desire/Goals/Goal.cs:64]: Maybe it's an idea to call UpdateStatus in the getter here before returning it?

File: Aplib.Core.Tests/Desire/GoalTests.cs [hunk 3]
```diff
@@ -124,12 +126,15 @@ public void Goal_WhereHeuristicsChange_UsesUpdatedHeuristics()
         // Arrange
         IBeliefSet beliefSetMock = Mock.Of<IBeliefSet>();
         bool shouldSucceed = false;
+        // ReSharper disable once AccessToModifiedClosure
```
Comment (JoenvandeVorle) [inline Aplib.Core.Tests/Desire/GoalTests.cs]: Reshaper leftover comment?

File: Aplib.Core.Tests/Desire/GoalTests.cs [hunk 4]
```diff
@@ -150,18 +155,68 @@ public void GoalConstructor_WhereHeuristicFunctionTypeDiffers_HasEqualBehaviour(
         // Arrange
         ITactic<IBeliefSet> tactic = Mock.Of<ITactic<IBeliefSet>>();
 
-        bool heuristicFunctionBoolean(IBeliefSet _) => goalCompleted;
+        // ReSharper disable once MoveLocalFunctionAfterJumpStatement
```
Comment (JoenvandeVorle) [inline Aplib.Core.Tests/Desire/GoalTests.cs]: same here

**Review by JoenvandeVorle — CHANGES_REQUESTED** [review]

File: Aplib.Core/Belief/Beliefs/Belief.cs (hunk 1 above)
Comment (Laura-Berkers) [reply Aplib.Core/Belief/Beliefs/Belief.cs:1]: It causes a naming collision between `System.Action` and `Aplib.Core.Intent.Actions.Action`. This has already caused som confusion amongst ourselves when writing code, because you might think you are creating or using an aplib Action but you are actually using a System.Action. This can also created very confusing error messages.

File: Aplib.Core/Desire/Goals/Goal.cs (hunk 2 above)
Comment (Laura-Berkers) [reply Aplib.Core/Desire/Goals/Goal.cs:64]: Yes that's a good idea actually.

Comment (Laura-Berkers) [reply Aplib.Core/Desire/Goals/Goal.cs:64]: I think actually we might remove the `UpdateStatus` method completely and just put it in the getter.

Comment (Laura-Berkers) [reply Aplib.Core/Desire/Goals/Goal.cs:64]: And also from goal structures, for consistency.

**Review by JensSteenmetz — APPROVED** [review]
LGTM!!

**Review by JoenvandeVorle — APPROVED** [review]
lgtm
